File: workloads/Python/bst/main-sampler.py
import sys
import argparse
import time
import json

from sampling import (
    sample_insert_valid,
    sample_delete_valid,
    sample_union_valid,
    sample_insert_post,
    sample_delete_post,
    sample_union_post,
    sample_insert_model,
    sample_delete_model,
    sample_union_model,
    sample_insert_insert,
    sample_insert_delete,
    sample_insert_union,
    sample_delete_insert,
    sample_delete_delete,
    sample_delete_union,
    sample_union_delete_insert,
    sample_union_union_idem,
    sample_union_union_assoc,
)

# Mapping from string names to property functions
PROPS = {
    "InsertValid": sample_insert_valid,
    "DeleteValid": sample_delete_valid,
    "UnionValid": sample_union_valid,
    "InsertPost": sample_insert_post,
    "DeletePost": sample_delete_post,
    "UnionPost": sample_union_post,
    "InsertModel": sample_insert_model,
    "DeleteModel": sample_delete_model,
    "UnionModel": sample_union_model,
    "InsertInsert": sample_insert_insert,
    "InsertDelete": sample_insert_delete,
    "InsertUnion": sample_insert_union,
    "DeleteInsert": sample_delete_insert,
    "DeleteDelete": sample_delete_delete,
    "DeleteUnion": sample_delete_union,
    "UnionDeleteInsert": sample_union_delete_insert,
    "UnionUnionIdem": sample_union_union_idem,
    "UnionUnionAssoc": sample_union_union_assoc,
}

class CustomEncoder(json.JSONEncoder):
    def default(self, o):
        if hasattr(o, "to_json"):
            return o.to_json()
        return super().default(o)

def main():
    parser = argparse.ArgumentParser(prog="rackcheck-bespoke")
    parser.add_argument("property", help="Name of the property to run")
    parser.add_argument("tests", type=int, help="Number of tests to run")
    args = parser.parse_args()

    if args.property not in PROPS:
        print(f"Unknown property: {args.property}")
        sys.exit(1)

    prop_fn = PROPS[args.property]

    print(f"Running {args.property} with {args.tests} tests...")

    # Run quick-sample using hypothesis.find()
    results = []
    for i in range(args.tests):
        t0 = time.time()
        e = prop_fn.strategy.example()
        t1 = time.time()
        duration = f"{t1 - t0:.6f}s"
        results.append({
            "time": duration,
            "value": repr(e),
        })
    print(json.dumps(results, cls=CustomEncoder))

if __name__ == "__main__":
    main()

File: workloads/OCaml/BST/bin/main_sampler.ml
open Cmdliner
open Sampler_qcheck
open BST.Gen_typebased_qcheck
open BST.Gen_bespoke_qcheck

let qcheck_property name =
  match name with
  | "InsertValid" -> prop_Q_InsertValid
  | "DeleteValid" -> prop_Q_DeleteValid
  | "UnionValid" -> prop_Q_UnionValid
  | "InsertPost" -> prop_Q_InsertPost
  | "DeletePost" -> prop_Q_DeletePost
  | "UnionPost" -> prop_Q_UnionPost
  | "InsertModel" -> prop_Q_InsertModel
  | "DeleteModel" -> prop_Q_DeleteModel
  | "UnionModel" -> prop_Q_UnionModel
  | "InsertInsert" -> prop_Q_InsertInsert
  | "InsertDelete" -> prop_Q_InsertDelete
  | "InsertUnion" -> prop_Q_InsertUnion
  | "DeleteUnion" -> prop_Q_DeleteUnion
  | "DeleteInsert" -> prop_Q_DeleteInsert
  | "DeleteDelete" -> prop_Q_DeleteDelete
  | "UnionDeleteInsert" -> prop_Q_UnionDeleteInsert
  | "UnionUnionIdem" -> prop_Q_UnionUnionIdem
  | "UnionUnionAssoc" -> prop_Q_UnionUnionAssoc
  | _ -> raise (Invalid_argument ("Unknown property: " ^ name))

let qcheck_generator name =
  match name with
  | "typebased" -> gen_Q_TypeBased
  | "bespoke" -> gen_Q_Bespoke
  | _ -> raise (Invalid_argument ("Unknown generator: " ^ name))

(** Command line setup *)

let main property strategy count seed =
  (* Your logic here: select property and generator by name *)
  let framework, generator =
    match String.split_on_char ':' strategy with
    | [ framework; generator ] -> (framework, generator)
    | _ -> failwith "Strategy must be in the form FRAMEWORK:GENERATOR"
  in
  match framework |> String.lowercase_ascii with
  | "qcheck" ->
      (qcheck_property property) ~gen:(qcheck_generator generator) ~count ~seed
  | _ -> failwith "sampling is only supported for qcheck framework"

(** *)

(** Cmdliner stuff *)

(** | *)

(** v *)

let _ = Random.self_init ()

let property_arg =
  let doc = "Name of the property test to run." in
  Arg.(
    required
    & opt (some string) None
    & info [ "property" ] ~docv:"PROPERTY" ~doc)

let generator_arg =
  let doc = "Name of the strategy to use." in
  Arg.(
    required
    & opt (some string) None
    & info [ "strategy" ] ~docv:"FRAMEWORK:GENERATOR" ~doc)

let count_arg =
  let doc = "Number of samples to generate." in
  Arg.(value & opt int 10 & info [ "count" ] ~docv:"COUNT" ~doc)

let seed_arg =
  let doc = "Random seed for the generator." in
  Arg.(value & opt int (Random.int 1000000) & info [ "seed" ] ~docv:"SEED" ~doc)

let term =
  Term.(const main $ property_arg $ generator_arg $ count_arg $ seed_arg)

let () = Cmd.(exit @@ eval (v (info "BST") term))
